src/typescript-worker/bare-module-transformer.ts

```diff
@@ -42,63 +122,293 @@ export class BareModuleTransformer {
+      // the preview iframe, because we're simply omitting this file from our
+      // output on error. We should instead allow FileBuildOutput to carry an
+      // HTTP status code, so then we could propagate this specific error to be
+      // served by the service worker, so that it shows up more usefully in the
+      // network tab.
+      console.error(`Error fetching ${path} from CDN: ${e.message}`);
+      return;
+    }
+    let packageJson: PackageJson | undefined | null = null;
+    const getPackageJson = async (): Promise<PackageJson | undefined> => {
+      if (packageJson === null) {
+        try {
+          packageJson = await this._cdn.fetchPackageJson(location);
+        } catch {
+          packageJson = undefined;
+        }
+      }
+      return packageJson;
+    };
+    yield* this._handleModule(
+      {
+        kind: 'file',
+        file: {
+          name: `node_modules/${path}`,
+          content: asset.content,
+          contentType: asset.contentType,
         },
       },
-    };
+      getPackageJson,
+      output
+    );
   }
 }
+
+/**
+ * Create a useful Playground diagnostic from an es-module-lexer exception.
+ */
+const makeEsModuleLexerDiagnostic = (
+  e: Error,
+  filename: string
+): DiagnosticBuildOutput | undefined => {
+  const match = e.message.match(/@:(\d+):(\d+)$/);
+  if (match === null) {
+    return undefined;
+  }
+  const line = Number(match[1]) - 1;
+  const character = Number(match[2]) - 1;
+  return {
+    kind: 'diagnostic',
+    filename,
+    diagnostic: {
+      message: `es-module-lexer error: ${e.message}`,
+      range: {
+        start: {line, character},
+        end: {line, character: character + 1},
+      },
+    },
+  };
+};
+
+/**
+ * Create a useful Playground diagnostic from a JSON.parse exception.
+ */
+const makeJsonParseDiagnostic = (
+  e: Error,
+  file: SampleFile
+): DiagnosticBuildOutput => {
+  const start = extractPositionFromJsonParseError(e.message, file.content) ?? {
+    line: 0,
+    character: 0,
+  };
+  return {
+    kind: 'diagnostic',
+    filename: file.name,
+    diagnostic: {
+      message: `Invalid package.json: ${e}`,
+      range: {
+        start,
+        // To the rest of the file.
+        end: charToLineAndChar(file.content, file.content.length),
+      },
+    },
+  };
+};
+
+/**
+ * Try to extract the line and character from a `JSON.parse` exception message.
+ *
+ * Chrome and Firefox often include this information, but using different
+ * formats. Safari never includes this information.
+ */
+const extractPositionFromJsonParseError = (
+  message: string,
+  json: string
+): {line: number; character: number} | undefined => {
+  const chrome = message.match(/at position (\d+)/);
+  if (chrome !== null) {
+    return charToLineAndChar(json, Number(chrome[1]));
+  }
+  const firefox = message.match(/at line (\d+) column (\d+)/);
+  if (firefox !== null) {
+    return {
+      line: Number(firefox[1]) - 1,
```

Comment:
Is there a difference between `Number(var)` and `parseInt(var, 10)`?

Reply:
Since we know we're only handling something that matches `\d+`, there's no difference in the output, and `Number` seemed shorter/simpler.

I like this table for seeing what all the differences are: https://stackoverflow.com/questions/17106681/parseint-vs-unary-plus-when-to-use-which/17106702#17106702